### PR TITLE
#256 JSON-RPC Package (& Commons)

### PR DESCRIPTION
**Pull request by MisterVants**
### Overview
This PR introduces a full compliant JSON-RPC 2.0 interface for request and responses. Specs can be found at [jsonrpc.org](https://www.jsonrpc.org/specification)

### What changed
- Introduces the `Commons` and `JSONRPC` packages.
- Adds `RPCRequest` and `RPCResponse` compliant with JSON-RPC 2.0 specifications.
- Adds `RPCRequestConvertible` to facilitate decouple of protocol methods from RPC requests.
- `AnyCodable` type moved to Commons package.
- Reintroduces `Either` type.
- Also enables the code coverage report for the project. (Was previously missing in SPM)

### Notes
Types were introduced in a way to interfere as little as possible with current SDK code. Current request objects can be migrated over time, and ID generation logic can be wrapped over the RPC models.

The JSONRPC types can be considered foundational code and are build with the intention to be flexible, solid, and allow use in other types of application. Feedback over the API and its usage is appreciated.

### CODE REVIEW

File: Sources/JSONRPC/RPCRequest.swift [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        id = try container.decodeIfPresent(RPCID.self, forKey: .id)
+        method = try container.decode(String.self, forKey: .method)
+        params = try container.decodeIfPresent(AnyCodable.self, forKey: .params)
+        if let decodedParams = params {
+            if decodedParams.value is Int || decodedParams.value is Double || decodedParams.value is String || decodedParams.value is Bool {
+                throw DecodingError.dataCorruptedError(
+                    forKey: .params,
+                    in: container,
+                    debugDescription: "")
```
Comment (MisterVants) [inline Sources/JSONRPC/RPCRequest.swift:107]: I forgot this description

File: Package.swift [hunk 2]
```diff
@@ -36,6 +36,12 @@ let package = Package(
             path: "Sources/WalletConnectKMS"),
         .target(
             name: "WalletConnectUtils",
+            dependencies: ["Commons"]),
```
Comment (llbartekll) [inline Package.swift:39]: Can we define a difference between `WalletConnectUtils` and a `Commons` ?

**Comment by llbartekll** [comment]
Looks, very clean and thoroughly tested, I just wonder if it is not too complex fro our needs but looking forward how it will work consumed by SDKs.

**Review by llbartekll — APPROVED** [review]

**Review by flypaper0 — APPROVED** [review]
Looks grate to me! Could you please commit a little example of integration JSON-RPC Package in out codebase?

**Comment by flypaper0** [comment]
We could close https://github.com/WalletConnect/WalletConnectSwiftV2/pull/155 after that changes?

**Comment by MisterVants** [comment]
> We could close #155 after that changes?

Absolutely

**Comment by MisterVants** [comment]
> Looks grate to me! Could you please commit a little example of integration JSON-RPC Package in out codebase?

Sure, I'll try doing something on a next PR

**Pull request merged**